### PR TITLE
Widget & weather pages

### PR DESCRIPTION
**Pull request by tkdeshan**

### CODE REVIEW

File: src/hooks/useFetch.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      if (error.response?.data?.code === 13004 && attempt <= 1) {
+        try {
+          await getNewAccessToken();
+          return callAPI();
+        } catch (refreshError) {
+          error = refreshError;
+
+          localStorage.removeItem('jwtAccessToken');
+          localStorage.removeItem('jwtRefreshToken');
+          navigate('/', { replace: true });
```
Comment (pankajan05) [inline src/hooks/useFetch.js]: replace '/; with '/login'.

File: src/hooks/useNotification.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  const { send } = useAxios();
+  const dispatch = useDispatch();
+  const userId = useSelector((state) => state.user._id);
+  const [notifications, setNotifications] = useState([]);
+  const [fetch, setNotificationFetch] = useState(false);
+
+  const fetchNotificationsData = async () => {
+    const res = await send({ endpoint: `notification/?userId=${userId}`, method: 'GET' });
+    setNotifications(res?.result || []);
+    dispatch(setNotificationCount(res?.result?.filter((data) => !data.readFlag).length || 0));
```
Comment (pankajan05) [inline src/hooks/useNotification.js:16]: make the fetch false once the notification fetched correctly.

File: src/pages/dashboard/pages/agro-eye-page/agro-eye-form/index.jsx [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    };
+
+    const response = await send({
+      endpoint: widget ? `widget/${widget._id}` : `widget`,
+      method: widget ? 'PUT' : 'POST',
+      body: requestData,
+    });
+
+    if (response) {
+      formSubmission(widget ? 'Widget updated successfully' : 'Widget created successfully');
```
Comment (pankajan05) [inline src/pages/dashboard/pages/agro-eye-page/agro-eye-form/index.jsx]: get these messages from constant file. Don't hard code the messages.

File: src/pages/dashboard/pages/agro-eye-page/chart/index.jsx [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+    const axisTime = []; // Generate time array between start and end time with time gap
+    let currentDateTime = new Date(startDateTime);
+    while (currentDateTime <= endDateTime) {
+      axisTime.push(new Date(currentDateTime).toISOString());
+      currentDateTime = new Date(currentDateTime.getTime() + timeGap);
+    }
+
+    const sensorSeries = widget?.devices.flatMap((device) => {
+      const deviceId = device.deviceId;
```
Comment (pankajan05) [inline src/pages/dashboard/pages/agro-eye-page/chart/index.jsx:39]: Too much variable creations . Is deviceId, factors necessary?

File: src/pages/dashboard/pages/agro-eye-page/chart/index.jsx [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        const seriesData = widget.sensorData
+          .find((sensorData) => sensorData[deviceId])
+          [deviceId].map((entry) => {
+            return {
+              x: new Date(entry.timestamp).toISOString(),
+              y: entry[factor],
+            };
+          });
+
+        const timeSeriesData = [];
```
Comment (pankajan05) [inline src/pages/dashboard/pages/agro-eye-page/chart/index.jsx:42]: move  all the variable timeSeriesData to starting of the function

File: src/pages/dashboard/pages/agro-eye-page/index.jsx [hunk 6]
```diff
@@ -1,184 +1,181 @@
...
+      const response = await send({
+        endpoint: `widget/delete/${selectedWidget._id}`,
+        method: 'PUT',
+        body: { isDeleted: true },
+      });
+      setIsConfirmVisible(false);
+      if (response) {
+        setMessage('Widget deleted successfully');
+        setIsAlertVisible(true);
+        recall();
```
Comment (pankajan05) [inline src/pages/dashboard/pages/agro-eye-page/index.jsx]: After deleted no need to fetch entire widget.

File: src/pages/dashboard/pages/weather-page/index.jsx [hunk 7]
```diff
@@ -1,31 +1,155 @@
...
 
 const WeatherComponent = () => {
+  const [location, setLocation] = useState(null);
+  const [weatherData, setWeatherData] = useState();
+  const [selectedDate, setSelectedDate] = useState(new Date());
+  const [btnIndex, setBtnIndex] = useState(0);
+  const dispatch = useDispatch();
+  const navigate = useNavigate();
+
+  const { response: user, isLoading } = useFetch({
```
Comment (pankajan05) [inline src/pages/dashboard/pages/weather-page/index.jsx]: get the user location from the user slice

**Review by pankajan05 — COMMENTED** [review]
checked

File: src/pages/dashboard/pages/agro-eye-page/chart/index.jsx (hunk 4 above)
Comment (tkdeshan) [reply src/pages/dashboard/pages/agro-eye-page/chart/index.jsx:39]: factors removed and simplify the function.
deviceId is used in 3 times

**Review by pankajan05 — APPROVED** [review]
Approved

**Pull request merged**